{"cells": [{"cell_type": "markdown", "metadata": {}, "source": ["# 1000 Genomes Chromosome 22 Variant Enrichment Analysis\n", "\n", "This notebook downloads chromosome 22 VCF data from the 1000 Genomes Project,\n", "extracts the first 10,000 variants, and performs a Chi-square test to evaluate\n", "if any variants are significantly enriched in allele frequency in one continental\n", "population versus others."]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["# Install necessary packages\n", "!pip install cyvcf2 pandas scipy tqdm --quiet"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["from cyvcf2 import VCF\n", "import pandas as pd\n", "from scipy.stats import chi2_contingency\n", "from tqdm import tqdm\n", "import os, gzip, shutil\n", "\n", "# Download VCF for chromosome 22\n", "vcf_url = \"ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\"\n", "vcf_file = \"ALL.chr22.phase3_shapeit2.vcf.gz\"\n", "\n", "if not os.path.exists(vcf_file):\n", "    !wget -O {vcf_file} {vcf_url}"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["# Load population metadata\n", "sample_info_url = \"ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_sample_info.txt\"\n", "sample_info_file = \"20130606_sample_info.txt\"\n", "\n", "if not os.path.exists(sample_info_file):\n", "    !wget -O {sample_info_file} {sample_info_url}\n", "\n", "sample_df = pd.read_csv(sample_info_file, sep='\\t')\n", "pop_map = sample_df.set_index('Sample')['Super Population'].to_dict()"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["# Initialize VCF parser\n", "vcf = VCF(vcf_file)\n", "samples = vcf.samples\n", "\n", "# Map samples to super populations\n", "sample_pops = [pop_map.get(s, None) for s in samples]\n", "pop_set = sorted(set([p for p in sample_pops if p]))\n", "\n", "# Count enriched variants\n", "from collections import defaultdict\n", "sig_counts = defaultdict(int)\n", "\n", "N = 10000\n", "for i, variant in enumerate(tqdm(vcf, total=N)):\n", "    if i >= N:\n", "        break\n", "    \n", "    genos = variant.genotypes  # List of tuples: (GT1, GT2, phased)\n", "    if not genos:\n", "        continue\n", "    \n", "    # Build contingency table\n", "    pop_alleles = {p: [0, 0] for p in pop_set}  # ALT, REF\n", "    for geno, pop in zip(genos, sample_pops):\n", "        if pop not in pop_alleles:\n", "            continue\n", "        alt_count = geno[0] + geno[1] if geno[0] != -1 and geno[1] != -1 else 0\n", "        ref_count = 2 - alt_count if alt_count <= 2 else 0\n", "        pop_alleles[pop][0] += alt_count\n", "        pop_alleles[pop][1] += ref_count\n", "\n", "    # Create 2xN table: rows = [ALT, REF]; columns = populations\n", "    table = [[], []]\n", "    for p in pop_set:\n", "        table[0].append(pop_alleles[p][0])\n", "        table[1].append(pop_alleles[p][1])\n", "\n", "    try:\n", "        chi2, pval, dof, ex = chi2_contingency(table)\n", "        if pval < 0.01:\n", "            for p in pop_set:\n", "                if pop_alleles[p][0] > sum([pop_alleles[x][0] for x in pop_set if x != p]) / (len(pop_set) - 1):\n", "                    sig_counts[p] += 1\n", "    except Exception:\n", "        continue"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["# Summary table\n", "summary_df = pd.DataFrame.from_dict(sig_counts, orient='index', columns=['Significant Variants'])\n", "summary_df = summary_df.sort_values(by='Significant Variants', ascending=False)\n", "summary_df"]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"name": "python", "version": "3.x"}}, "nbformat": 4, "nbformat_minor": 5}